File: Code/Asteroid_Trajectory_Selection/COMPUTE_PATH_DELTAV.m
function [delta_v_launch, delta_v_A1_arrive, delta_v_A1_leave, ...
    delta_v_A2_arrive, delta_v_A2_leave, delta_v_A3_arrive, delta_v_total] ...
    = COMPUTE_PATH_DELTAV(A_ID_1, A_ID_2, A_ID_3, et_launch, ...
    et_arrive_1, et_stay_1, et_arrive_2, et_stay_2, et_arrive_3, M_1, M_2, M_3)

    MINUTE = 60;
    HOUR = 60 * MINUTE;
    DAY = 24 * HOUR;

    EARTH_LAUNCH_STATE = cspice_spkezr('399', et_launch, 'ECLIPJ2000', 'NONE', '10');
    A_1_ARRIVE_STATE = cspice_spkezr(A_ID_1, et_arrive_1, 'ECLIPJ2000', 'NONE', '10');
    A_1_LEAVING_STATE = cspice_spkezr(A_ID_1, et_stay_1, 'ECLIPJ2000', 'NONE', '10');
    A_2_ARRIVE_STATE = cspice_spkezr(A_ID_2, et_arrive_2, 'ECLIPJ2000', 'NONE', '10');
    A_2_LEAVING_STATE = cspice_spkezr(A_ID_2, et_stay_2, 'ECLIPJ2000', 'NONE', '10');
    A_3_ARRIVE_STATE = cspice_spkezr(A_ID_3, et_arrive_3, 'ECLIPJ2000', 'NONE', '10');

    MYU_SUN = cspice_bodvcd(10, 'GM', 10);

    [EARTH_LAMBERT_VELOCITY, A_1_ARRIVE_LAMBERT_VELOCITY, ~, EXIT_FLAG_1] = ...
        lambert(EARTH_LAUNCH_STATE(1:3)', A_1_ARRIVE_STATE(1:3)', ...
        (et_arrive_1 - et_launch) / DAY, M_1, MYU_SUN);
    [A_1_LEAVE_LAMBERT_VELOCITY, A_2_ARRIVE_LAMBERT_VELOCITY, ~, EXIT_FLAG_2] = ...
        lambert(A_1_LEAVING_STATE(1:3)', A_2_ARRIVE_STATE(1:3)', ...
        (et_arrive_2 - et_stay_1) / DAY, M_2, MYU_SUN);
    [A_2_LEAVE_LAMBERT_VELOCITY, A_3_ARRIVE_LAMBERT_VELOCITY, ~, EXIT_FLAG_3] = ...
        lambert(A_2_LEAVING_STATE(1:3)', A_3_ARRIVE_STATE(1:3)', ...
        (et_arrive_3 - et_stay_2) / DAY, M_3, MYU_SUN);

    if EXIT_FLAG_1 ~= 1 || EXIT_FLAG_2 ~= 1 || EXIT_FLAG_3 ~= 1
        fprintf("LAMBERT SOLVER FAILED TO CONVERGE FOR ASTEROIDS WITH IDS %s, %s, %s.\n", ...
            A_ID_1, A_ID_2, A_ID_3);
        delta_v_launch = -1;
        delta_v_A1_arrive = -1;
        delta_v_A1_leave = -1;
        delta_v_A2_arrive = -1;
        delta_v_A2_leave = -1;
        delta_v_A3_arrive = -1;
        delta_v_total = -1;
        return
        
    end

    delta_v_launch = EARTH_LAMBERT_VELOCITY' - EARTH_LAUNCH_STATE(4:6);
    delta_v_A1_arrive = A_1_ARRIVE_LAMBERT_VELOCITY' - A_1_ARRIVE_STATE(4:6);
    delta_v_A1_leave = A_1_LEAVE_LAMBERT_VELOCITY' - A_1_LEAVING_STATE(4:6);
    delta_v_A2_arrive = A_2_ARRIVE_LAMBERT_VELOCITY' - A_2_ARRIVE_STATE(4:6);
    delta_v_A2_leave = A_2_LEAVE_LAMBERT_VELOCITY' - A_2_LEAVING_STATE(4:6);
    delta_v_A3_arrive = A_3_ARRIVE_LAMBERT_VELOCITY' - A_3_ARRIVE_STATE(4:6);
    

    delta_v_total = norm(delta_v_launch) + norm(delta_v_A1_arrive) + ...
        norm(delta_v_A1_leave) + norm(delta_v_A2_arrive) + ...
        norm(delta_v_A2_leave) + norm(delta_v_A3_arrive);



end

File: Python/asteroid_trajectory_selection/compute_path_deltav.py
import numpy as np
import spiceypy

from .constants import DAY
from .lambert import lambert


def compute_path_deltav(a_id_1, a_id_2, a_id_3, et_launch,
                        et_arrive_1, et_stay_1, et_arrive_2,
                        et_stay_2, et_arrive_3, m_1, m_2, m_3):
    earth_launch_state, _ = spiceypy.spkezr('399', et_launch, 'ECLIPJ2000', 'NONE', '10')
    a_1_arrive_state, _ = spiceypy.spkezr(str(a_id_1), et_arrive_1, 'ECLIPJ2000', 'NONE', '10')
    a_1_leaving_state, _ = spiceypy.spkezr(str(a_id_1), et_stay_1, 'ECLIPJ2000', 'NONE', '10')
    a_2_arrive_state, _ = spiceypy.spkezr(str(a_id_2), et_arrive_2, 'ECLIPJ2000', 'NONE', '10')
    a_2_leaving_state, _ = spiceypy.spkezr(str(a_id_2), et_stay_2, 'ECLIPJ2000', 'NONE', '10')
    a_3_arrive_state, _ = spiceypy.spkezr(str(a_id_3), et_arrive_3, 'ECLIPJ2000', 'NONE', '10')

    _, myu_sun_vals = spiceypy.bodvcd(10, 'GM', 10)
    myu_sun = myu_sun_vals[0]

    earth_lambert_velocity, a_1_arrive_lambert_velocity, _, exit_flag_1 = \
        lambert(earth_launch_state[0:3], a_1_arrive_state[0:3],
                (et_arrive_1 - et_launch) / DAY, m_1, myu_sun)
    a_1_leave_lambert_velocity, a_2_arrive_lambert_velocity, _, exit_flag_2 = \
        lambert(a_1_leaving_state[0:3], a_2_arrive_state[0:3],
                (et_arrive_2 - et_stay_1) / DAY, m_2, myu_sun)
    a_2_leave_lambert_velocity, a_3_arrive_lambert_velocity, _, exit_flag_3 = \
        lambert(a_2_leaving_state[0:3], a_3_arrive_state[0:3],
                (et_arrive_3 - et_stay_2) / DAY, m_3, myu_sun)

    if exit_flag_1 != 1 or exit_flag_2 != 1 or exit_flag_3 != 1:
        return {
            'delta_v_launch': np.array([]),
            'delta_v_A1_arrive': np.array([]),
            'delta_v_A1_leave': np.array([]),
            'delta_v_A2_arrive': np.array([]),
            'delta_v_A2_leave': np.array([]),
            'delta_v_A3_arrive': np.array([]),
            'delta_v_total': 1e3,
        }

    delta_v_launch = earth_lambert_velocity - earth_launch_state[3:6]
    delta_v_A1_arrive = a_1_arrive_lambert_velocity - a_1_arrive_state[3:6]
    delta_v_A1_leave = a_1_leave_lambert_velocity - a_1_leaving_state[3:6]
    delta_v_A2_arrive = a_2_arrive_lambert_velocity - a_2_arrive_state[3:6]
    delta_v_A2_leave = a_2_leave_lambert_velocity - a_2_leaving_state[3:6]
    delta_v_A3_arrive = a_3_arrive_lambert_velocity - a_3_arrive_state[3:6]

    delta_v_total = (np.linalg.norm(delta_v_A1_arrive)
                     + np.linalg.norm(delta_v_A1_leave)
                     + np.linalg.norm(delta_v_A2_arrive)
                     + np.linalg.norm(delta_v_A2_leave)
                     + np.linalg.norm(delta_v_A3_arrive))

    return {
        'delta_v_launch': delta_v_launch,
        'delta_v_A1_arrive': delta_v_A1_arrive,
        'delta_v_A1_leave': delta_v_A1_leave,
        'delta_v_A2_arrive': delta_v_A2_arrive,
        'delta_v_A2_leave': delta_v_A2_leave,
        'delta_v_A3_arrive': delta_v_A3_arrive,
        'delta_v_total': delta_v_total,
    }
